File: api/get-analytics.py
import json
import os
from datetime import datetime
from http.server import BaseHTTPRequestHandler
from urllib.request import Request, urlopen

KV_URL = os.environ.get('KV_REST_API_URL')
KV_TOKEN = os.environ.get('KV_REST_API_TOKEN')

def kv_call(path):
    if not KV_URL or not KV_TOKEN:
        raise RuntimeError("KV not configured")
    req = Request(f"{KV_URL}{path}")
    req.add_header("Authorization", f"Bearer {KV_TOKEN}")
    with urlopen(req) as resp:
        return json.loads(resp.read().decode())

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        try:
            # Load last 10k views from KV
            try:
                res = kv_call("/LRANGE/views/0/9999")
                items = res.get('result', [])
                views = [json.loads(s) for s in items]
            except Exception:
                views = []

            analytics = self.process_analytics(views)

            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
            self.send_header('Access-Control-Allow-Headers', 'Content-Type')
            self.end_headers()
            self.wfile.write(json.dumps(analytics).encode())

        except Exception as e:
            print(f"Error getting analytics: {str(e)}")
            self.send_response(500)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({'error': 'Internal server error'}).encode())

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def process_analytics(self, views):
        page_stats, unique_users, daily_stats = {}, set(), {}
        for view in views:
            page = view['page']
            if page not in page_stats:
                page_stats[page] = {
                    'totalViews': 0,
                    'uniqueUsers': set(),
                    'firstView': view['timestamp'],
                    'lastView': view['timestamp']
                }
            page_stats[page]['totalViews'] += 1
            page_stats[page]['uniqueUsers'].add(view['sessionId'])

            t0 = datetime.fromisoformat(page_stats[page]['firstView'].replace('Z', '+00:00'))
            t1 = datetime.fromisoformat(page_stats[page]['lastView'].replace('Z', '+00:00'))
            vt = datetime.fromisoformat(view['timestamp'].replace('Z', '+00:00'))
            if vt < t0:
                page_stats[page]['firstView'] = view['timestamp']
            if vt > t1:
                page_stats[page]['lastView'] = view['timestamp']

            unique_users.add(view['sessionId'])

            date = view['date']
            if date not in daily_stats:
                daily_stats[date] = {'views': 0, 'users': set()}
            daily_stats[date]['views'] += 1
            daily_stats[date]['users'].add(view['sessionId'])

        formatted_page_stats = {
            p: {
                'totalViews': s['totalViews'],
                'uniqueUsers': len(s['uniqueUsers']),
                'firstView': s['firstView'],
                'lastView': s['lastView']
            }
            for p, s in page_stats.items()
        }
        formatted_daily_stats = {
            d: { 'views': s['views'], 'uniqueUsers': len(s['users']) }
            for d, s in daily_stats.items()
        }
        return {
            'pages': formatted_page_stats,
            'dailyStats': formatted_daily_stats,
            'totals': {
                'totalViews': len(views),
                'totalUniqueUsers': len(unique_users),
                'totalPages': len(page_stats)
            },
            'lastUpdated': datetime.now().isoformat()
        }

File: api/get-analytics.js
// Vercel API endpoint: /api/get-analytics
// This endpoint returns analytics data for the admin dashboard

import fs from 'fs';
import path from 'path';

export default async function handler(req, res) {
  // Only allow GET requests
  if (req.method !== 'GET') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  try {
    // Read views data
    const dataPath = path.join(process.cwd(), 'data', 'views.json');
    let views = [];

    try {
      if (fs.existsSync(dataPath)) {
        const fileContent = fs.readFileSync(dataPath, 'utf8');
        views = JSON.parse(fileContent);
      }
    } catch (error) {
      console.error('Error reading views file:', error);
      return res.status(500).json({ error: 'Failed to read analytics data' });
    }

    // Process analytics data
    const analytics = processAnalytics(views);

    res.status(200).json(analytics);

  } catch (error) {
    console.error('Error getting analytics:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
}

function processAnalytics(views) {
  // Group by page
  const pageStats = {};
  const uniqueUsers = new Set();
  const dailyStats = {};

  views.forEach(view => {
    const page = view.page;
    
    // Initialize page stats
    if (!pageStats[page]) {
      pageStats[page] = {
        totalViews: 0,
        uniqueUsers: new Set(),
        firstView: view.timestamp,
        lastView: view.timestamp
      };
    }

    // Update page stats
    pageStats[page].totalViews++;
    pageStats[page].uniqueUsers.add(view.sessionId);
    
    // Update first and last view timestamps
    const currentFirst = new Date(pageStats[page].firstView);
    const currentLast = new Date(pageStats[page].lastView);
    const viewTime = new Date(view.timestamp);
    
    if (viewTime < currentFirst) {
      pageStats[page].firstView = view.timestamp;
    }
    if (viewTime > currentLast) {
      pageStats[page].lastView = view.timestamp;
    }

    // Track unique users globally
    uniqueUsers.add(view.sessionId);

    // Daily stats
    const date = view.date;
    if (!dailyStats[date]) {
      dailyStats[date] = { views: 0, users: new Set() };
    }
    dailyStats[date].views++;
    dailyStats[date].users.add(view.sessionId);
  });

  // Convert Sets to counts and format data
  const formattedPageStats = {};
  Object.keys(pageStats).forEach(page => {
    formattedPageStats[page] = {
      totalViews: pageStats[page].totalViews,
      uniqueUsers: pageStats[page].uniqueUsers.size,
      firstView: pageStats[page].firstView,
      lastView: pageStats[page].lastView
    };
  });

  // Format daily stats
  const formattedDailyStats = {};
  Object.keys(dailyStats).forEach(date => {
    formattedDailyStats[date] = {
      views: dailyStats[date].views,
      uniqueUsers: dailyStats[date].users.size
    };
  });

  // Calculate totals
  const totalViews = views.length;
  const totalUniqueUsers = uniqueUsers.size;

  return {
    pages: formattedPageStats,
    dailyStats: formattedDailyStats,
    totals: {
      totalViews,
      totalUniqueUsers,
      totalPages: Object.keys(pageStats).length
    },
    lastUpdated: new Date().toISOString()
  };
}
